source/extensions/clusters/aggregate/cluster.h

```diff
@@ -0,0 +1,158 @@
+
+class Cluster : public Upstream::ClusterImplBase, Upstream::ClusterUpdateCallbacks {
+public:
+  Cluster(const envoy::api::v2::Cluster& cluster,
+          const envoy::config::cluster::aggregate::ClusterConfig& config,
+          Upstream::ClusterManager& cluster_manager, Runtime::Loader& runtime,
+          Runtime::RandomGenerator& random,
+          Server::Configuration::TransportSocketFactoryContext& factory_context,
+          Stats::ScopePtr&& stats_scope, ThreadLocal::SlotAllocator& tls, bool added_via_api);
+
+  // Upstream::Cluster
+  Upstream::Cluster::InitializePhase initializePhase() const override {
+    return Upstream::Cluster::InitializePhase::Secondary;
+  }
+
+  // Upstream::ClusterUpdateCallbacks
+  void onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) override;
+  void onClusterRemoval(const std::string& cluster_name) override;
+
+  void refresh() {
+    refresh([](const std::string&) { return false; });
+  }
+
+  std::vector<std::string> clusters_;
+  Upstream::ClusterUpdateCallbacksHandlePtr handle_;
+  Upstream::ClusterManager& cluster_manager_;
+  Runtime::Loader& runtime_;
+  Runtime::RandomGenerator& random_;
+  ThreadLocal::SlotPtr tls_;
+
+private:
+  // Upstream::ClusterImplBase
+  void startPreInit() override;
+
+  void refresh(const std::function<bool(const std::string&)>& skip_predicate);
+  PriorityContext
+  linearizePrioritySet(const std::function<bool(const std::string&)>& skip_predicate);
+};
+
+// Load balancer used by each worker thread. It will be refreshed when clusters, hosts or priorities
+// are updated.
+class AggregateClusterLoadBalancer : public Upstream::LoadBalancer {
+public:
+  AggregateClusterLoadBalancer(Upstream::ClusterStats& stats, Runtime::Loader& runtime,
+                               Runtime::RandomGenerator& random,
+                               const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+      : stats_(stats), runtime_(runtime), random_(random), common_config_(common_config) {}
+
+  // Upstream::LoadBalancer
+  Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override;
+
+private:
+  // Use inner class to extend LoadBalancerBase. When initializing AggregateClusterLoadBalancer, the
+  // priority set could be empty, we cannot initialize LoadBalancerBase when priority set is empty.
+  class LoadBalancerImpl : public Upstream::LoadBalancerBase {
+  public:
+    LoadBalancerImpl(const PriorityContext& priority_context, Upstream::ClusterStats& stats,
+                     Runtime::Loader& runtime, Runtime::RandomGenerator& random,
+                     const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+        : Upstream::LoadBalancerBase(priority_context.first, stats, runtime, random, common_config),
+          priority_to_cluster_(priority_context.second) {}
+
+    // Upstream::LoadBalancer
+    Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override;
```

Comment:
No. I think we can override either `chooseHost` or `chooseHostOnce`. Perhaps, overriding `chooseHostOnce` is better as it will respect the retry count and other retry policies.

Reply:
Discussed offline. Keep current implementation. If overriding `chooseHostOnce`, it could trigger multiple retries on multiple levels.